### ISSUE TITLE
Non-stopping containers corrupt internal state with  "pod rm -f" 

### ISSUE THREAD

**Issue by ensc** [containers/libpod#4704]
/kind bug

**Description**

When a container does not get stopped by `podman pod rm -f` information about it are not completely released and e.g. its name can not be reused anymore.  Such containers do not show up in "podman ps -a" and can not be removed.  Their`slurp4netns` network stays alive too and blocks reuse of mapped ports.

**Steps to reproduce the issue:**

1.  `podman pod create --name test1`

2.  add a container which needs some time for shutdown; e.g. `postgres` 

     ```
     podman run -d --pod test1 --name c1 postgres
     ```

3.  remove the pod forcefully and wait

     ```
     podman pod rm -f test1
     ```

     After some time you will see "cannot remove container ... did not die within timeout"

4.  repeat step 1 + 2

**Describe the results you received:**

Step 4.2 errors with

```
Error: error creating container storage: the container name "c1" is already in use by "ffa8ebb5bc118a5e3b5b872ba7eb76edd471d74bb8049a840f844d6828e1f0a2". You have to remove that container to be able to reuse that name.: that name is already in use
```

Removing it does not work either:

```
$ podman rm -f c1
Error: Failed to evict container: "": Failed to find container "c1" in state: no container with name or ID c1 found: no such container
$ podman rm -f ffa8ebb5bc118a5e3b5b872ba7eb76edd471d74bb8049a840f844d6828e1f0a2
Error: Failed to evict container: "": Failed to find container "ffa8ebb5bc118a5e3b5b872ba7eb76edd471d74bb8049a840f844d6828e1f0a2" in state: no container with name or ID ffa8ebb5bc118a5e3b5b872ba7eb76edd471d74bb8049a840f844d6828e1f0a2 found: no such container
```

A `podman ps -a` shows a running infra container but not `c1`.  

**Describe the results you expected:**

Name `c1` can be used again

**Output of `podman version`:**

```
$ podman version
Version:            1.6.2
RemoteAPI Version:  1
Go Version:         go1.13.1
OS/Arch:            linux/amd64
```

**Output of `podman info --debug`:**

```
debug:
  compiler: gc
  git commit: ""
  go version: go1.13.1
  podman version: 1.6.2
host:
  BuildahVersion: 1.11.3
  CgroupVersion: v2
  Conmon:
    package: conmon-2.0.2-1.fc31.x86_64
    path: /usr/bin/conmon
    version: 'conmon version 2.0.2, commit: 186a550ba0866ce799d74006dab97969a2107979'
  Distribution:
    distribution: fedora
    version: "31"
  IDMappings:
    gidmap:
    - container_id: 0
      host_id: 505
      size: 1
    - container_id: 1
      host_id: 505000000
      size: 1000000
    uidmap:
    - container_id: 0
      host_id: 505
      size: 1
    - container_id: 1
      host_id: 505000000  
      size: 1000000
  MemFree: 18385846272
  MemTotal: 32669814784   
  OCIRuntime:
    name: crun
    package: crun-0.10.6-1.fc31.x86_64
    path: /usr/bin/crun   
    version: |-
      crun version 0.10.6 
      spec: 1.0.0
      +SYSTEMD +SELINUX +APPARMOR +CAP +SECCOMP +EBPF +YAJL
  SwapFree: 3845648384
  SwapTotal: 4294963200   
  arch: amd64
  cpus: 8
  eventlogger: journald   
  hostname: ensc-virt.intern.sigma-chemnitz.de
  kernel: 5.3.15-300.fc31.x86_64
  os: linux
  rootless: true
  slirp4netns:
    Executable: /usr/bin/slirp4netns
    Package: slirp4netns-0.4.0-20.1.dev.gitbbd6f25.fc31.x86_64
    Version: |-
      slirp4netns version 0.4.0-beta.3+dev
      commit: bbd6f25c70d5db2a1cd3bfb0416a8db99a75ed7e
  uptime: 93h 20m 33.35s (Approximately 3.88 days)
registries:
  blocked: null
  insecure: null
  search:
  - docker.io
  - registry.fedoraproject.org
  - registry.access.redhat.com
  - registry.centos.org   
  - quay.io
store:
  ConfigFile: /home/ensc/.config/containers/storage.conf
  ContainerStore:
    number: 15
  GraphDriverName: overlay
  GraphOptions:
    overlay.mount_program:
      Executable: /usr/bin/fuse-overlayfs
      Package: fuse-overlayfs-0.7.2-2.fc31.x86_64
      Version: |-
        fusermount3 version: 3.6.2
        fuse-overlayfs: version 0.7.2
        FUSE library version 3.6.2
        using FUSE kernel interface version 7.29
  GraphRoot: /.local/home/ensc/.local/share/containers/storage
  GraphStatus:
    Backing Filesystem: extfs
    Native Overlay Diff: "false"
    Supports d_type: "true"
    Using metacopy: "false"
  ImageStore:
    number: 20
  RunRoot: /run/user/505  
  VolumePath: /.local/home/ensc/.local/share/containers/storage/volumes
```

**Package info (e.g. output of `rpm -q podman` or `apt list podman`):**

```
podman-1.6.2-2.fc31.x86_64
```

**Comment by mheon** [comment]
Will be fixed in 1.7.0

**Comment by ensc** [comment]
fwiw, problem still exists in

```
$ podman --version
podman version 1.7.0-rc1

$ rpm -q podman
podman-1.7.0-0.2.rc1.fc31.x86_64
```

**Comment by mheon** [comment]
I assume you're running rootless?

Can you provide the full error message (`cannot remove container ... did not die within timeout`)? This could be the container not stopping after SIGKILL, which is somewhat concerning.

**Comment by ensc** [comment]
yes; it is rootless.

```
[ensc@ensc-virt ~]$ podman pod create --name test1
be8613c444097da06c80164a907beb1a703e090b378597c5713a4e774d96751d
[ensc@ensc-virt ~]$ podman run -d --pod test1 --name c1 postgres
5885530ddaf5a183c539fa75cd614e668a416d6755e59a3289252abfeeda30a9
[ensc@ensc-virt ~]$ podman pod rm -f test1
Error: cannot remove container 5885530ddaf5a183c539fa75cd614e668a416d6755e59a3289252abfeeda30a9 as it could not be stopped: given PIDs did not die within timeout
[ensc@ensc-virt ~]$ podman pod create --name test1
8220e428bd7f4134e7309365a651a5404230dc543ed5d25b8e4c76c83b8526a6
[ensc@ensc-virt ~]$ podman run -d --pod test1 --name c1 postgres
Error: error creating container storage: the container name "c1" is already in use by "5885530ddaf5a183c539fa75cd614e668a416d6755e59a3289252abfeeda30a9". You have to remove that container to be able to reuse that name.: that name is already in use
[ensc@ensc-virt ~]$ podman --version
podman version 1.7.0-rc1
```

**Comment by Komic** [comment]
Also seeing this, not rootless.

**Comment by mheon** [comment]
I can't reproduce here. I assume there's some sort of condition here causing Postgres to not die?

Can you check if the container PID is still there after the container dies? I wonder if you're seeing a lot of zombie processes stuck in I/O sleep, or similar. I don't think we can do much about that - if you have processes that won't die and are using the mount, we can't remove it.

**Comment by Komic** [comment]
As far as I can tell no, nothing is running. Getting `Error: container is mounted and cannot be removed without using force: container state improper` when trying to rm --storage

**Comment by mheon** [comment]
And you're certain you're getting `did not die within timeout` errors?

Are you using `podman rm` or `podman pod rm`

**Comment by Komic** [comment]
Pretty sure yes. Both pod rm and rm fail but the actual container is what I seem to have issues with, not the infra one.

**Comment by ensc** [comment]
Here, it might be caused by an old `~/.config/containers/libpod.conf` configuration which had

```
cgroup_manager = "cgroupfs"
```

After removing this configuration, a new one with

```
cgroup_manager = "systemd"
```

will be created where the problem does not happen anymore.

I am pretty sure that `cgroupfs` was not added by me but some former setting.

**Comment by Komic** [comment]
Afraid I don't have such a file on my system.

**Comment by rhatdan** [comment]
We have another issue with podman rm -f versus podman stop/podman rm.  The issue is we are setting the "ALL" flag to true in podman rm case, where as in podman stop we are setting ALL to false.

The ALL flag tells the OCI Runtime to send the signal to ALL processes versus just PID 1 of the container.  If the Container is running in a PID Namespace, then sending the signal to PID 1 is enough, since when it dies all of the processes of the container die. 
crun has a bug that is failing when calling it with the ALL flag, which is being worked on by @giuseppe 

Bottom line, I believe podman rm -f should not set the ALL Flag unless it is in the Host pid namespace.

This might fix this issue.

**Comment by mheon** [comment]
The changes related to the work @rhatdan was talking about should be in 1.7, so this may be fixed

**Comment by Komic** [comment]
I think I'm still having related issues on 1.7.

```
# podman pod rm -f example-pod
2020-01-18T15:31:22.000354839Z: error opening file '/run/crun/0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a/status': No such file or directory
ERRO[0005] Error removing container 0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a from pod 972aa81ed9a97925fe58ddf61c3ed3d6fd937bc62d92aa040d562ff2a939fcc8: cannot remove container 0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a as it could not be stopped: timed out waiting for file /var/run/libpod/exits/0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a: internal libpod error
2020-01-18T15:31:27.000370881Z: error opening file '/run/crun/97f38e8bca4fa49450b2d8eac7fcd103c67bfcfbabce7543d10aa3edd3 151adc/status': No such file or directory
ERRO[0010] Error removing container 97f38e8bca4fa49450b2d8eac7fcd103c67bfcfbabce7543d10aa3edd3151adc from pod 972aa81ed9a97925fe58ddf61c3ed3d6fd937bc62d92aa040d562ff2a939fcc8: cannot remove container 97f38e8bca4fa49450b2d8eac7fcd103c67bfcfbabce7543d10aa3edd3151adc as it could not be stopped: timed out waiting for file /var/run/libpod/exits/97f38e8bca4fa49450b2d8eac7fcd103c67bfcfbabce7543d10aa3edd3151adc: internal libpod error
972aa81ed9a97925fe58ddf61c3ed3d6fd937bc62d92aa040d562ff2a939fcc8

# podman play kube /home/example/pod-example.yml
Error: error creating container storage: the container name "example" is already in use by "0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a". You have to remove that container to be able to reuse that name.: that name is already in use

# podman rm -f example
Error: Failed to evict container: "": Failed to find container "example" in state: no container with name or ID example found: no such container
```

A combination of `rm -fv` and `--storage` on `example` and `0f77bf6fb52dd06c240877df081a7e156383c1030e6550593c5c052013d69a3a` allows me to play kube the container again.

**Comment by rhatdan** [comment]
@Komic I think you are seeing a different issue, please open a new issue.
